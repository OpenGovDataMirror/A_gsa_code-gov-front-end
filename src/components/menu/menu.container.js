/* global PUBLIC_PATH */

import { withRouter } from 'react-router-dom'
import { connect } from 'react-redux'
import { getConfigValue, isHomepage } from 'utils/other'
import toggleSearchDropdown from 'actions/toggle-search-dropdown'
import MenuComponent from './menu.component'

export const mapStateToProps = ({ router, searchDropdown }) => ({
    color: 'white',
    logoDark: getConfigValue('content.header.logos.dark'),
    logoLight: getConfigValue('content.header.logos.light'),
    menu: getConfigValue('content.header.menu'),
    onHomePage: isHomepage,
    searchDropdown,
    siteTitle: getConfigValue('title')
})

export const mapDispatchToProps = dispatch => ({
  toggleSearchDropdown: () => dispatch(toggleSearchDropdown())
})

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(MenuComponent)
)
