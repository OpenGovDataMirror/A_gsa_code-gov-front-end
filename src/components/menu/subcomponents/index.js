import PrimaryMenuOption from './primary-menu-option'
import SecondaryDropdown from './secondary-dropdown'
import SearchBoxDropDown from './search-box-dropdown'

export { PrimaryMenuOption, SearchBoxDropDown, SecondaryDropdown }
