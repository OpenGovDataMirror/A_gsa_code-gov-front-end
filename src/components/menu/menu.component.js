/* global PUBLIC_PATH */

import React, { Component, Fragment } from 'react'
import CustomLink from 'components/custom-link'
import { map } from '@code.gov/cautious'
import MobileMenuSearchBoxComponent from 'components/mobile-menu-search-box'
import { PrimaryMenuOption, SecondaryDropdown } from './subcomponents'

export default class Menu extends Component {
  constructor(props) {
    super(props)
    this.state = {
      menu: props.menu,
      mobileMenu: false
    }

    this.onToggleMenuOption = this.onToggleMenuOption.bind(this)
    this.onToggleMobileMenu = this.onToggleMobileMenu.bind(this)
    this.header = React.createRef()
  }

  onToggleMobileMenu(event) {
    const mobileMenu = !this.state.mobileMenu
    this.setState({ mobileMenu })
  }

  onToggleMenuOption(selected, event) {
    const menu = this.state.menu.map(menuOption => {
      if (menuOption.expanded && menuOption.name !== selected.name) {
        menuOption.expanded = false
      }
      if (menuOption.name === selected.name) {
        selected.expanded = !selected.expanded
        return selected
      }
      return menuOption
    })
    this.setState({ menu })
  }

  get menus() {
    return this.state.menu.map(menuOption => (
      <Fragment key={menuOption.name}>
        <PrimaryMenuOption menuOption={menuOption} onClick={this.onToggleMenuOption} />
        <SecondaryDropdown menuOption={menuOption} />
      </Fragment>
    ))
  }

  collapse() {
    const menu = this.state.menu.map(menuOption => {
      menuOption.expanded = false
      return menuOption
    })

    this.setState({
      menu
    })

    const mobileMenu = false
    this.setState({ mobileMenu })
  }

  render() {
    const { color, siteTitle } = this.props
    return (
      <>
        <div className="usa-overlay" />
        <header className="usa-header usa-header--basic bg-white">
          <div className="usa-nav-container">
            <div className="usa-navbar border-bottom-0">
              <div className="usa-logo" id="basic-logo">
                <CustomLink to="/" className="svg-container code-logo" title={`${siteTitle} Home`}>
                  <img
                    src={color === 'white' ? this.props.logoDark : this.props.logoLight}
                    alt="code.gov"
                    className="height-4 maxw-card"
                    aria-label="Home"
                  />
                </CustomLink>
              </div>
              <button
                className="usa-menu-btn bg-white"
                onClick={::this.onToggleMobileMenu}
                aria-labelledby="open-menu"
              >
                <i
                  className="icon icon-menu text-primary font-body-xl"
                  role="img"
                  aria-label="open menu"
                  id="open-menu"
                />
              </button>
            </div>
            <nav
              aria-label="Primary navigation"
              className={this.state.mobileMenu ? 'usa-nav is-visible' : 'usa-nav'}
            >
              <button
                className="usa-nav__close"
                onClick={::this.onToggleMobileMenu}
                aria-labelledby="close-menu"
              >
                <i
                  className="icon icon-close text-primary font-body-lg padding-right-4 padding-top-2"
                  role="img"
                  aria-label="close menu"
                  id="close-menu"
                />
              </button>
              <ul className="usa-nav__primary usa-accordion">
                {map(this.state.menu, menuOption => (
                  <li className="usa-nav__primary-item" key={menuOption.name}>
                    <PrimaryMenuOption
                      menuOption={menuOption}
                      onClick={::this.onToggleMenuOption}
                      onSelectLink={::this.onToggleMobileMenu}
                    />
                    <SecondaryDropdown menuOption={menuOption} onClick={::this.collapse} />
                  </li>
                ))}
              </ul>
            </nav>
          </div>
        </header>
      </>
    )
  }
}
