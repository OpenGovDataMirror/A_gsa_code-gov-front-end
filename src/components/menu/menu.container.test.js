import toggleSearchDropdown from 'actions/toggle-search-dropdown'
import { mapStateToProps, mapDispatchToProps } from 'components/menu/menu.container'

jest.mock('actions/toggle-search-dropdown')

const props = {
  router: {
    location: {
      pathname: '/test-path'
    }
  },
  searchDropdown: 'test-search-dropdown'
}

const dispatch = jest.fn()

describe('containers - Menu', () => {
  describe('mapStateToProps', () => {
    it('should return the correct properties', () => {
      expect(mapStateToProps(props)).toMatchSnapshot()
    })

    it('should set the `color` to `white` if on the home page', () => {
      const router = { location: { pathname: PUBLIC_PATH } }
      expect(mapStateToProps({ ...props, router }).color).toBe('white')
    })
  })

  describe('mapDispatchToProps', () => {
    describe('toggleSearchDropdown', () => {
      it('should dispatch the `toggleSearchDropdown` action', () => {
        mapDispatchToProps(dispatch).toggleSearchDropdown()
        expect(dispatch).toBeCalled()
        expect(toggleSearchDropdown).toBeCalled()
      })
    })
  })
})
