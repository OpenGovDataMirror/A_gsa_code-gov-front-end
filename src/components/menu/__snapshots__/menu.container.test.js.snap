// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers - Menu mapStateToProps should return the correct properties 1`] = `
Object {
  "color": "white",
  "logoDark": "/assets/img/logos/code-gov-dark.svg",
  "logoLight": "/assets/img/logos/code-gov-light.svg",
  "menu": Array [
    Object {
      "name": "AGENCIES",
      "url": "/agencies",
    },
    Object {
      "links": Array [
        Object {
          "name": "Agency Compliance",
          "url": "/agency-compliance/compliance/dashboard",
        },
        Object {
          "name": "How to Procure Software",
          "url": "/agency-compliance/compliance/procurement",
        },
        Object {
          "name": "How to Inventory Code",
          "url": "/agency-compliance/compliance/inventory-code",
        },
      ],
      "name": "GUIDANCE",
    },
  ],
  "onHomePage": true,
  "searchDropdown": "test-search-dropdown",
  "siteTitle": "code.gov",
}
`;
